src/gprs.c

```diff
@@ -3186,21 +3192,49 @@ static void gprs_load_settings(struct ofono_gprs *gprs, const char *imsi)
 static struct pri_context *gprs_context_for_ia(struct ofono_gprs *gprs)
 {
 	GSList *l;
+	struct pri_context *ctx_ia = NULL;
+	struct pri_context *ctx_inet_pref = NULL;
+	struct pri_context *ctx_inet = NULL;
+	struct pri_context *ctx_other_pref = NULL;
 
 	for (l = gprs->contexts; l; l = l->next) {
 		struct pri_context *ctx = l->data;
 
-		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA)
-			return ctx;
-	}
+		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_IA) {
+			if (ctx->preferred)
+				return ctx;
 
-	for (l = gprs->contexts; l; l = l->next) {
-		struct pri_context *ctx = l->data;
+			if (ctx_ia == NULL)
+				ctx_ia = ctx;
+		} else if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_INTERNET) {
 
-		if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_INTERNET)
-			return ctx;
+			if (ctx->preferred && ctx_inet_pref == NULL)
+				ctx_inet_pref = ctx;
+			else if (ctx_inet == NULL)
+				ctx_inet = ctx;
+
+		} else if (ctx->preferred && ctx_other_pref == NULL) {
+			ctx_other_pref = ctx;
+		}
+	}
+
+	if (ctx_ia != NULL) {
+		set_preferred(ctx_ia, ofono_dbus_get_connection(), TRUE);
+		return ctx_ia;
 	}
 
+	if (ctx_inet_pref != NULL)
+		return ctx_inet_pref;
+
+	if (ctx_inet != NULL)
+		return ctx_inet;
+
+	if (ctx_other_pref != NULL)
+		return ctx_other_pref;
+
+	if (gprs->contexts == NULL)
+		return NULL;
+
 	return gprs->contexts->data;
```

Comment:
The NULL check really needs to be before this final return.  If (gprs->contexts == NULL), it'll crash...